servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,372 @@
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.netty.util.internal.PlatformDependent.throwException;
+import static io.servicetalk.buffer.api.ReadOnlyBufferAllocators.DEFAULT_RO_ALLOCATOR;
+import static io.servicetalk.buffer.netty.BufferUtils.getByteBufAllocator;
+import static io.servicetalk.buffer.netty.BufferUtils.newBufferFrom;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static java.util.Objects.requireNonNull;
+
+final class NettyChannelContentCodec extends AbstractContentCodec {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyChannelContentCodec.class);
+    private static final Buffer END_OF_STREAM = DEFAULT_RO_ALLOCATOR.fromAscii(" ");
+    private static final int MAX_SIZE_FOR_MERGED_BUFFER = 1024; //1KiB
+
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier;
+    private final Supplier<ByteToMessageDecoder> decoderSupplier;
+
+    NettyChannelContentCodec(final CharSequence name,
+                             final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier,
+                             final Supplier<ByteToMessageDecoder> decoderSupplier) {
+        super(name);
+        this.encoderSupplier = encoderSupplier;
+        this.decoderSupplier = decoderSupplier;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        requireNonNull(allocator);
+
+        if (src.readableBytes() == 0) {
+            return src;
+        }
+
+        if (offset < 0 || offset >= src.readableBytes()) {
+            throw new IllegalArgumentException("Invalid offset: " + offset +
+                    " (expected 0 - " + (src.readableBytes() - 1) + ")");
+        }
+
+        final int availableBytes = src.readableBytes() - offset;
+        if (length > availableBytes || length < 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected 0 - " + availableBytes + ")");
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeOutbound(origin);
+
+            // May produce footer
+            preparePendingData(channel);
+
+            final Buffer buffer = drainChannelQueueToSingleBuffer(channel.outboundMessages(), allocator);
+            if (buffer == null) {
+                throw new CodecEncodingException(this, "Not enough data to produce an encoded output");
+            }
+
+            cleanup(channel);
+            return buffer;
+        } catch (CodecEncodingException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new CodecEncodingException(this, e);
+        } finally {
+            safeCleanup(channel);
```

Comment:
Ok, I see the picture now. Let's keep the `finally` block. Thanks!
Ignore my previous https://github.com/apple/servicetalk/pull/1376#discussion_r592504997